proposals/XXXX-efficient-token-program.md

```diff
@@ -0,0 +1,97 @@
+
+1. [`withdraw_excess_lamports`](https://github.com/solana-program/token/blob/main/p-token/src/processor/withdraw_excess_lamports.rs) (instruction discriminator `38`): allow recovering "bricked" SOL from mint (e.g., USDC mint as `~323` SOL in excess) and multisig accounts. The logic of this instruction is the same as the current SPL Token-2022 instruction. 
+2. [`batch`](https://github.com/solana-program/token/blob/main/p-token/src/processor/batch.rs) (instruction discriminator `255`): enable efficient CPI interaction with the Token program. This is a new instruction that can execute a variable number on Token instructions in a single invocation of the Token program. Therefore, the CPI invoke units (currently `1000` CU) are only consumed once, instead of each CPI instruction – this significantly improves the CUs required to perform multiple Token instructions in a CPI context.
+
+Note that `withdraw_excess_lamports` discriminator matches the value used in SPL Token-2022, while `batch` has a driscriminator value that is not used in either SPL Token nor Token-2022.
+
+The program will be loaded into an account (`ptokNfvuU7terQ2r2452RzVXB3o4GT33yPWo1fUkkZ2`) prior to enabling the feature gate that triggers the replacement.
+
+When the feature gate `ptokSWRqZz5u2xdqMdstkMKpFurauUpVen7TZXgDpkQ` is enabled, the runtime needs to:
+
+* Replace `TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA` program with `ptokNfvuU7terQ2r2452RzVXB3o4GT33yPWo1fUkkZ2` using the Upgradable Loader `v4`.
+
+Another point for consideration is to whether to keep the current logs of the program or not. Logs are limited to show the name of the instruction being executed, e.g.:
+```
+Program TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA invoke [1]
+Program log: Instruction: Transfer
+Program TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA consumed 249 of 200000 compute units
+Program TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA success
+```
+Logging the instruction name consumes around `103` compute units, which in some cases reprensents almost the same amount required to execute the instruction. Note that errors will always be logged, since this is considered good use of CU.
+
+## Alternatives Considered
+
+As an alternative to replace the current version of SPL Token, `p-token` could be deployed to a new address and people can be encouraged to transition to that program. This would hinder its adoption and benefits, since people would be very slow to adopt the new program, if they adopt it at all.
+
+## Impact
+
+The main impact is freeing up block CUs, allowing more transactions to be packed in a block; dapp developers benefit since interacting with the Token program will consume significantly less CUs.
+
+Below is a sample of the CU efficient gained by `p-token` compared to the current SPL Token program.
+
+| Instruction                | CU (`p-token`) | CU (`p-token`) + logging | CU (`spl-token`) |
```

Comment:
Can you add the percent of CUs used for each instruction? For example, for `InitializeMint`, under `p-token`, specify (4%) and under `p-token` + logging specify (7%)? That might make it clearer just how costly the logging is. I'm open to suggestions of how we can present this more clearly